addons/mail/models/mail_thread.py

```diff
@@ -4123,30 +4132,61 @@ def _truncate_payload(self, payload):
-                return False
+        if isinstance(partner_ids, (list, tuple)):
+            _logger.warning("message_subscribe now expects a dict mapping res_id:partner_ids instead of a simple partner_ids list", stack_info=True)
+            if not self:
+                return True
+            mapping = {record.id: partner_ids for record in self}
         else:
-            self.check_access_rights('write')
-            self.check_access_rule('write')
+            mapping = {record_id: pids or [] for record_id, pids in partner_ids.items()}
 
         # filter inactive and private addresses
-        if partner_ids and not adding_current:
-            partner_ids = self.env['res.partner'].sudo().search([('id', 'in', partner_ids), ('active', '=', True)]).ids
+        all_partner_ids = list(set(itertools.chain.from_iterable(mapping.values())))
+        all_partner_ids = self.env['res.partner'].sudo().search([('id', 'in', all_partner_ids), ('active', '=', True)]).ids
+        all_partner_set = set(all_partner_ids)
+
+        record_ids_to_check_read = []
+        record_ids_to_check_write = []
+
+        mapping_adding_current = {}
+        mapping_not_adding_current = {}
+
+        current_user_partner_set = set([self.env.user.partner_id.id])
+        for record_id, pids in mapping.items():
+            adding_current = set(pids) == current_user_partner_set
+            pids = list(all_partner_set.intersection(pids))
+            if adding_current:
+                mapping_adding_current[record_id] = pids
+            else:
+                mapping_not_adding_current[record_id] = pids
 
-        return self._message_subscribe(partner_ids, subtype_ids, customer_ids=customer_ids)
+            if pids and adding_current:
+                record_ids_to_check_read.append(record_id)
+            else:
+                record_ids_to_check_write.append(record_id)
 
-    def _message_subscribe(self, partner_ids=None, subtype_ids=None, customer_ids=None):
+        if record_ids_to_check_read:
+            try:
+                records_to_check_read = self.env[self._name].browse(record_ids_to_check_read)
+                records_to_check_read.check_access_rights('read')
+                records_to_check_read.check_access_rule('read')
+            except exceptions.AccessError:
+                return False
+        if record_ids_to_check_write:
+            records_to_check_write = self.env[self._name].browse(record_ids_to_check_write)
+            records_to_check_write.check_access_rights('write')
+            records_to_check_write.check_access_rule('write')
+
+        if mapping_adding_current:
+            self._message_subscribe(mapping_adding_current, subtype_ids, customer_ids=[])
+        if mapping_not_adding_current:
+            self._message_subscribe(mapping_not_adding_current, subtype_ids, customer_ids=None)
+        return True
+
+    def _message_subscribe(self, partners_mapping=None, subtype_ids=None, customer_ids=None):
```

Comment:
there is `def _message_subscribe` in /home/odoo/odoo/odoo/addons/mail/models/discuss/discuss_channel.py
Probably API should be changed there as well